packages/util/src/signature.ts

```diff
@@ -18,30 +31,56 @@ export interface ECDSASignature {
  * accordingly, otherwise return a "static" `v` just derived from the `recovery` bit
  */
 export function ecsign(msgHash: Buffer, privateKey: Buffer, chainId?: bigint): ECDSASignature {
-  const [signature, recovery] = signSync(msgHash, privateKey, { recovered: true, der: false })
-
+  const [signature, rec] = signSync(msgHash, privateKey, { recovered: true, der: false })
   const r = Buffer.from(signature.slice(0, 32))
   const s = Buffer.from(signature.slice(32, 64))
-
-  const v =
-    chainId === undefined
-      ? BigInt(recovery + 27)
-      : BigInt(recovery + 35) + BigInt(chainId) * BigInt(2)
-
-  return { r, s, v }
-}
-
-function calculateSigRecovery(v: bigint, chainId?: bigint): bigint {
-  if (v === BigInt(0) || v === BigInt(1)) return v
-
-  if (chainId === undefined) {
-    return v - BigInt(27)
+  const recovery = BigInt(rec)
+  let v
+  if (chainId !== undefined) {
+    v = recovery + BigInt(35) + chainId * BigInt(2)
+  } else {
+    v = BigInt(rec + 27)
   }
-  return v - (chainId * BigInt(2) + BigInt(35))
+  return { r, s, v, recovery }
 }
+/**
+ * Based on the solving recovery formula
+ * v - (chainId * BigInt(2) + BigInt(35))
+ * To an integer or non-integer
+ * Resulting in recovery (yParity) value of 0 or 1.
+ * Which determines which point on the EC is used
+ */
 
-function isValidSigRecovery(recovery: bigint): boolean {
-  return recovery === BigInt(0) || recovery === BigInt(1)
+export function calculateSigRecovery(v: bigint, chainId?: bigint): bigint {
+  if (chainId !== undefined) {
+    if (v === BigInt(0) || v === BigInt(1)) {
+      return v
+    } else if (v === BigInt(27) || v === BigInt(28)) {
+      return v - BigInt(27)
+    } else {
+      return v - (chainId * BigInt(2) + BigInt(35))
+    }
+  } else {
+    if (v === BigInt(0) || v === BigInt(1)) {
+      return v
+    } else if (v < BigInt(27)) {
+      throw new Error('invlaid v value < 27')
+    } else if (v > BigInt(28) && v < BigInt(35)) {
+      throw new Error('invlaid v value >28 & <35 ')
+    } else if (v > BigInt(36)) {
+      throw new Error('invlaid v value > 36')
```

Comment:
Also can you put minimally the error cases for the *generally*  invalid v values out of this conditional here and put at the beginning of the function code, since these are generic and apply for all cases. So this would be:

- `v > BigInt(1) && v < BigInt(27)` and
- `v > BigInt(28) && v < BigInt(35)`

(correct me if I am not correct here or missing something, also writing this down somewhat quickly)

I would also be a fan if you compatify the check to one conditional for all values and then just give one generic error message just passing in the actual v value, so `Invalid signature v value ${v}`.